x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -221,22 +231,32 @@ protected void masterOperation(
                 return;
             }
 
+            final MapperService mapperService = indicesService.createIndexMapperServiceForValidation(sourceIndexMetadata);
+            final CompressedXContent sourceIndexCompressedXContent = new CompressedXContent(sourceIndexMappings);
+            mapperService.merge(MapperService.SINGLE_MAPPING_NAME, sourceIndexCompressedXContent, MapperService.MergeReason.INDEX_TEMPLATE);
+
             final String mapping;
             try {
-                mapping = createRollupIndexMapping(request.getRollupConfig(), dimensionFieldCaps, metricFieldCaps);
+                mapping = createRollupIndexMapping(
+                    helper,
+                    request.getRollupConfig(),
+                    mapperService,
+                    sourceIndexMappings,
+                    listener::onFailure
+                );
             } catch (IOException e) {
                 listener.onFailure(e);
                 return;
             }
```

Comment:
I see that you pass the failure handler `listener:onFailure` all the way down to the `addMetricFields()` method.
Also, you catch the `IOException` and call the failure handler here. 

I think you should choose one of the two approaches here. Either throw exception in the calling methods and wrap it in the failure handler at the top level or use the failure handler wherever an exception is thrown. 

Since methods `createRollupIndexMapping()` and `addMetricFields()` are called synchronously, you can signal errors by throwing exceptions. The exceptions will be called and delegated to the failure handler at the top level. I think this will simplify testing the methods signatures a bit and make testing easier.